### PR TITLE
feat: add Docker & Docker Compose

### PR DESCRIPTION
**Pull request by geekyharsh05**
This pr setups:
- [x] Docker multistage build for smaller image size of `168MB`.

Resolves: #8 

### All services are running corectly
Screenshot 2024-10-08 at 4 06 13 AM

### Prisma Studio
Screenshot 2024-10-08 at 4 23 03 AM

### Next app
Screenshot 2024-10-08 at 4 23 05 AM

@anisharaz Can you please review it?

### CODE REVIEW

**Comment by geekyharsh05** [comment]
@anisharaz These secrets are only for the docker dev mode. So that is why i hard coded them.

File: .env.exampe
Comment (anisharaz) [inline .env.exampe:1]: The file is duplicate

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,57 @@
+FROM node:20-alpine AS base
+
+# Stage 1: Install dependencies
+FROM base AS installer
+
+WORKDIR /usr/src/app
+RUN apk add --no-cache libc6-compat
+COPY package.json package-lock.json ./
+COPY prisma ./prisma
```
Comment (anisharaz) [inline Dockerfile:9]: Copy of prisma is not necessery in this stage, please remove and it should be added in the next stage

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    else \
+    echo "package-lock.json not found" && exit 1; \
+    fi
+
+# Stage 2: Build stage
+FROM base AS builder
+WORKDIR /usr/src/app
+
+COPY --from=installer /usr/src/app/node_modules ./node_modules
+COPY --from=installer /usr/src/app/prisma ./prisma
```
Comment (anisharaz) [inline Dockerfile:23]: replace this line 23 with `copy . .`  which will copy all file along with prisma in this stage

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+RUN DATABASE_URL=$DATABASE_URL npx prisma generate
+
+COPY . .
```
Comment (anisharaz) [inline Dockerfile:27]: It can be removed from here since we are performing `copy . .` in upper lines

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    fi
+
+# Stage 2: Build stage
+FROM base AS builder
+WORKDIR /usr/src/app
+
+COPY --from=installer /usr/src/app/node_modules ./node_modules
+COPY --from=installer /usr/src/app/prisma ./prisma
+
+RUN DATABASE_URL=$DATABASE_URL npx prisma generate
```
Comment (anisharaz) [inline Dockerfile:25]: providing database url is not necessary for prisma generate. You can add `npx prisma generate` in the same line when you run `npm run build`

**Review by anisharaz — COMMENTED** [review]
@geekyharsh05 i appreciate you efforts for adding the commit. As described please can you remove all other changes other than `Dockerfile`.

**Comment by geekyharsh05** [comment]
@anisharaz Should I remove the dev file and the compose file?

**Comment by anisharaz** [comment]
yes absolutely

File: Dockerfile (hunk 2 above)
Comment (geekyharsh05) [reply Dockerfile:23]: But will it not be better to copy only prisma file?

Comment (anisharaz) [reply Dockerfile:23]: it is copying prisma file from previous stage which doesnot uses prisma file. so instead we can directly copy all file along with prisma file in this stage.

**Comment by geekyharsh05** [comment]
@anisharaz

**Comment by anisharaz** [comment]
Thankyou @geekyharsh05 for the contribution.

**Comment by geekyharsh05** [comment]
@anisharaz Thank you!

**Pull request merged**